### PR TITLE
Updated assumptions: Now complex and !finite infers infinite.

### LINKED ISSUE #16579: What should be the definition of infinite?
Consider the following code,
```python
>>> from sympy import symbols
>>> x = symbols('x', real=True, finite=False)
>>> x.is_infinite  # None
>>>
```
Isn't `real and not finite => infinite`?

### PR DESCRIPTION
**Pull request by ShubhamKJha**
#### References to other Issues or PRs

Fixes #16579 

#### Brief description of what is fixed or changed
A complex value can either be `finite` or `infinite`. Therefore, `complex and not finite` should mean the value is `infinite`. Currently, SymPy doesn't account for this. 
With this PR,
```py
>>> from sympy import Dummy
>>> Dummy(complex=True, finite=False).is_infinite
True
```

#### Release Notes

* core
  * added new clause `complex  ->   finite | infinite` in `_assume_rules` in assumptions.py.

### CODE REVIEW

**Comment by jksuom** [comment]
Please add a test.

**Comment by oscargus** [comment]
I'm not that well versed with the assumption system, will this also mean that the same thing hold for e.g. `real`? 

(Wouldn't it be enough to add an assumption as `finite == !infinite`? Which I realize that there already is one. So this is really the connection that any number that is `finite` is not `infinite` and vice versa? Maybe the release note should be updated to clarify this?)

**Comment by ShubhamKJha** [comment]
HI @oscargus we can't take `finite == !infinite`, there can numbers which are both not finite and not infinite (e.g. `nan`). The case that `finite infers not infinite` holds for `complex` numbers (`real` numbers are `complex` so they are covered).
How should I update release notes?

**Comment by oscargus** [comment]
OK! Yeah, now I noted that it is only in one direction, so infinite -> !finite, not the other way around. 

I'm just thinking that maybe one should focus on the practical effect in the release note. Now, if you specify any property of the number (complex, imaginary, real, rational, integer) and either finite or infinite, it will inherit the other finite/infinite property. This is clearly not how it should be formulated though... But more the practical consequence than the code change.

This also means that specifying e.g. real means that it cannot take on NaN by definition and that is really the only difference between that and not specifying anything.

It leads to the question: what does this really say about `x`? 

`Symbol('x', finite=False, infinite=False)`

`x`is NaN?

**Comment by ShubhamKJha** [comment]
> It leads to the question: what does this really say about `x`?
> 
> `Symbol('x', finite=False, infinite=False)`
> 
> `x`is NaN?

Now I see, `nan.is_finite` and `nan.is_infinite` both return `None`. Though clearly there can be consequences of having `finite  ==  !infinite`, I can't find an example which breaks this. Also see this [commit](https://github.com/sympy/sympy/issues/16579#issuecomment-480586015)'s message.

**Comment by asmeurer** [comment]
I guess I originally wanted to add `complex -> finite` in https://github.com/sympy/sympy/pull/8293, but it was too hard. This is a good step forward. 

We should decide once and for all however exactly which predicates should contain infinity.

File: sympy/core/tests/test_assumptions.py [hunk 1]
```diff
@@ -1062,3 +1062,10 @@ def test_issue_16313():
     assert (l*x).is_real is False
     assert (l*x*x).is_real is None  # since x*x can be a real number
     assert (-x).is_positive is False
+
+def test_issue_16579():
```
Comment (asmeurer) [inline sympy/core/tests/test_assumptions.py]: Maybe add a comment here that complex may not imply infinite in the future.

Comment (ShubhamKJha) [reply sympy/core/tests/test_assumptions.py]: Okay, I would add comment.

**Comment by ShubhamKJha** [comment]
@jksuom is it good to be merged?

**Comment by jksuom** [comment]
> With this complex and not finite would infer infinite. Since reals are complex, the implication follows in real as well.

I would remove this explanation from the release notes. Maybe it, or part of it, could go in the comments above. Otherwise I think this is good to be merged.

**Comment by ShubhamKJha** [comment]
Hi @jksuom considering the work in #16603 the scenario with reals may change so I am not adding this to comments. I revised the release notes please merge it.

**Comment by jksuom** [comment]
Thanks.

**Pull request merged**